pkg/ccl/crosscluster/logical/logical_replication_job_test.go

```diff
@@ -1031,6 +1031,152 @@ func TestHeartbeatCancel(t *testing.T) {
+	defer log.Scope(t).Close(t)
+
+	ctx := context.Background()
+
+	clusterArgs := base.TestClusterArgs{
+		ServerArgs: base.TestServerArgs{
+			DefaultTestTenant: base.TestControlsTenantsExplicitly,
+			Knobs: base.TestingKnobs{
+				JobsTestingKnobs: jobs.NewTestingKnobsWithShortIntervals(),
+				DistSQL:          &execinfra.TestingKnobs{},
+			},
+		},
+	}
+
+	server, s, runners, dbNames := setupServerWithNumDBs(t, ctx, clusterArgs, 1, 3)
+	defer server.Stopper().Stop(ctx)
+
+	PGURLs, cleanup := GetPGURLs(t, s, dbNames)
+	defer cleanup()
+
+	dbA, dbB, dbC := runners[0], runners[1], runners[2]
+
+	var (
+		jobAID jobspb.JobID
+		jobBID jobspb.JobID
+	)
+	dbC.QueryRow(t, "CREATE LOGICAL REPLICATION STREAM FROM TABLE tab ON $1 INTO TABLE tab", PGURLs[0].String()).Scan(&jobAID)
+	dbC.QueryRow(t, "CREATE LOGICAL REPLICATION STREAM FROM TABLE tab ON $1 INTO TABLE tab", PGURLs[1].String()).Scan(&jobBID)
+
+	// Insert into dest, then check source2 -> dest wins
+	dbC.Exec(t, "UPSERT INTO tab VALUES (1, 'hello')")
+	dbB.Exec(t, "UPSERT INTO tab VALUES (1, 'goodbye')")
+	now := s.Clock().Now()
+	WaitUntilReplicatedTime(t, now, dbC, jobBID)
+	expectedRows := [][]string{
+		{"1", "goodbye"},
+	}
+	dbC.CheckQueryResults(t, "SELECT * from tab", expectedRows)
+
+	// Write to source1 and source2, which should keep their respective rows but dest should resolve a conflict
+	dbA.Exec(t, "UPSERT INTO tab VALUES (1, 'insertA')")
+	dbB.Exec(t, "UPSERT INTO tab VALUES (1, 'insertB')")
+
+	expectedRowsS1 := [][]string{
+		{"1", "insertA"},
+	}
+	expectedRowsDest := [][]string{
+		{"1", "insertB"},
+	}
+	now = s.Clock().Now()
+	WaitUntilReplicatedTime(t, now, dbC, jobAID)
+	WaitUntilReplicatedTime(t, now, dbC, jobBID)
+	dbA.CheckQueryResults(t, "SELECT * from tab", expectedRowsS1)
+	dbB.CheckQueryResults(t, "SELECT * from tab", expectedRowsDest)
+	dbC.CheckQueryResults(t, "SELECT * from tab", expectedRowsDest)
+}
+
+// TestFullyConnectedReplication tests 4 tables that are all streaming
+// from each other and how they handle conflicts
+func TestFullyConnectedReplication(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	skip.UnderRace(t, "Replication doesn't complete in time")
+
+	ctx := context.Background()
+
+	clusterArgs := base.TestClusterArgs{
+		ServerArgs: base.TestServerArgs{
+			DefaultTestTenant: base.TestControlsTenantsExplicitly,
+			Knobs: base.TestingKnobs{
+				JobsTestingKnobs: jobs.NewTestingKnobsWithShortIntervals(),
+			},
+		},
+	}
+
+	verifyExpectedRowAllServers := func(
+		t *testing.T, runners []*sqlutils.SQLRunner, expectedRows [][]string, dbNames []string,
+	) {
+		for i, name := range dbNames {
+			runners[i].CheckQueryResults(t, fmt.Sprintf("SELECT * from %s.tab", name), expectedRows)
+		}
+	}
+
+	waitUntilReplicatedTimeAllServers := func(
+		t *testing.T,
+		targetTime hlc.Timestamp,
+		runners []*sqlutils.SQLRunner,
+		jobIDs [][]jobspb.JobID,
+	) {
+		for destIdx, srcJobIDs := range jobIDs {
```

Comment:
srcJobIDs is confusing as it suggests you're iterating over the jobs on the source cluster, which is not the case. I think it would be cleaner to iterate like so:

```
for destIdx := range jobIDs {
   for srcIdx := range jobIDs[destIdx] {
      waitUntilReplicatedTime( ..... jobIDs[destIdx][srcIdx[)
  }
}
```